src/shm.h

```diff
@@ -0,0 +1,571 @@
+    }
+};
+
+#endif
+
+template<typename T>
+struct SharedMemoryBackendFallback {
+    SharedMemoryBackendFallback() = default;
+
+    SharedMemoryBackendFallback(const std::string&, const T& value) :
+        fallback_object(make_unique_large_page<T>(value)) {}
+
+    void* get() const { return fallback_object.get(); }
+
+    SharedMemoryBackendFallback(const SharedMemoryBackendFallback&)            = delete;
+    SharedMemoryBackendFallback& operator=(const SharedMemoryBackendFallback&) = delete;
+
+    SharedMemoryBackendFallback(SharedMemoryBackendFallback&& other) noexcept :
+        fallback_object(std::move(other.fallback_object)) {}
+
+    SharedMemoryBackendFallback& operator=(SharedMemoryBackendFallback&& other) noexcept {
+        fallback_object = std::move(other.fallback_object);
+        return *this;
+    }
+
+   private:
+    LargePagePtr<T> fallback_object;
+};
+
+
+// Platform-independent wrapper
+template<typename T>
+struct SystemWideSharedConstant {
+   private:
+    static std::string createHashString(const std::string& input) {
+        size_t hash = std::hash<std::string>{}(input);
+
+        std::stringstream ss;
+        ss << std::hex << std::setfill('0') << hash;
+
+        return ss.str();
+    }
+
+   public:
+    // We can't run the destructor because it may be in a completely different process.
+    // The object stored must also be obviously in-line but we can't check for that, other than some basic checks that cover most cases.
+    static_assert(std::is_trivially_destructible_v<T>);
+    static_assert(std::is_trivially_move_constructible_v<T>);
+    static_assert(std::is_trivially_copy_constructible_v<T>);
+
+    SystemWideSharedConstant() = default;
+
+
+    // Content is addressed by its hash. An additional discriminator can be added to account for differences
+    // that are not present in the content, for example NUMA node allocation.
+    SystemWideSharedConstant(const T& value, std::size_t discriminator = 0) {
+        std::size_t content_hash    = std::hash<T>{}(value);
+        std::size_t executable_hash = std::hash<std::string>{}(getExecutablePathHash());
+
+        std::string shm_name = std::string("Local\\") + std::to_string(content_hash) + "$"
+                             + std::to_string(executable_hash) + "$"
+                             + std::to_string(discriminator);
+
+#ifndef _WIN32
+        // POSIX shared memory names must start with a slash
+        shm_name = "/" + createHashString(shm_name);
+
+        // hash name and make sure it is not longer than NAME_MAX
+        if (shm_name.size() > NAME_MAX)
+        {
+            shm_name = shm_name.substr(0, NAME_MAX);
+        }
+#endif
+
+        SharedMemoryBackend<T> shm_backend(shm_name, value);
+
+        if (shm_backend.is_valid())
+        {
+            backend = std::move(shm_backend);
+        }
+        else
+        {
+            std::cerr << "Failed to create shared memory backend: "
```

Comment:
probably this should not be written to stderr, especially since it is not a hard error, and will be invisible to most people using a GUI or similar, but maybe use UCIEngine::print_info_string ? (a bit like we do in the verify networks)

Reply:
Coupling it with UCI sounds bad, and doing it in a different way would be convoluted. I agree that it's not ideal, but potentially not that important to tackle at this point.

Reply:
I meant for the networks we pass a lambda doing
```
uci.cpp:    engine.set_on_verify_networks([](const auto& s) { print_info_string(s); });
uci.cpp:    engine.set_on_verify_networks([](const auto&) {});
```
which is roughly what I meant.

Reply:
I think to achieve this we have to remove the printout here and instead check the status after loading the network. Will explore this.